utils/controller/controller.go

```diff
@@ -18,6 +23,43 @@ import (
 	logutil "github.com/argoproj/argo-rollouts/utils/log"
 )
 
+// processNextWatchObj will process a single object from the watch by seeing if
+// that object is in an index and enqueueing the value object from the indexer
+func processNextWatchObj(watchEvent watch.Event, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) {
+	obj := watchEvent.Object
+	acc, err := meta.Accessor(obj)
+	if err != nil {
+		log.Errorf("Error processing object from watch: %v", err)
+		return
+	}
+	objsToEnqueue, err := indexer.ByIndex(index, fmt.Sprintf("%s/%s", acc.GetNamespace(), acc.GetName()))
+	if err != nil {
+		log.Errorf("Cannot process indexer: %s", err.Error())
+		return
+	}
+	for i := range objsToEnqueue {
+		Enqueue(objsToEnqueue[i], queue)
+	}
+}
+
+// WatchResource is a long-running function that will continually call the
+// processNextWatchObj function in order to watch changes on a resource kind
+// and enqueue a different resources kind that interact with them
+func WatchResource(client dynamic.Interface, gvk schema.GroupVersionResource, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) {
+	log.Infof("Starting watch on resource '%s'", gvk.Resource)
+	watch, err := client.Resource(gvk).Watch(metav1.ListOptions{})
+	if k8serrors.IsNotFound(err) {
+		log.Infof("Watch for resource '%s' failed: resource not found", gvk.Resource)
+		return
+	}
+	if err != nil {
+		panic(err)
```

Comment:
If we panic here, does this stop the `wait.Until()` goroutine in the caller? 

I think WatchResource() might be better off surfacing errors so that the caller can decide how to handle the various errors.